backend/src/database/migrations/20220915123536-create-email-messages-tx.js

```diff
@@ -0,0 +1,87 @@
+'use strict'
+
+const { TransactionalEmailMessageStatus } = require('@email/models')
+
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    await queryInterface.createTable('email_messages_tx', {
+      id: {
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.BIGINT,
+      },
+      user_email: {
+        type: Sequelize.DataTypes.TEXT,
+        allowNull: false,
+        references: {
+          model: 'users',
+          key: 'email',
+        },
+        onUpdate: 'CASCADE',
+      },
+      from_name: {
+        allowNull: false,
+        type: Sequelize.DataTypes.STRING(255),
+      },
+      from_address: {
+        allowNull: false,
+        type: Sequelize.DataTypes.STRING(255),
+      },
+      recipient: {
+        allowNull: false,
+        type: Sequelize.DataTypes.STRING(255),
+      },
+      params: {
+        type: Sequelize.DataTypes.JSON,
+        allowNull: false,
+      },
+      has_attachment: {
+        type: Sequelize.DataTypes.BOOLEAN,
+        allowNull: false,
+      },
+      attachment_s3_object: {
+        type: Sequelize.DataTypes.JSON,
+        allowNull: true,
+      },
+      status: {
+        type: Sequelize.DataTypes.ENUM(
+          TransactionalEmailMessageStatus.Unsent,
+          TransactionalEmailMessageStatus.InvalidFromAddressError,
+          TransactionalEmailMessageStatus.RateLimitError,
+          TransactionalEmailMessageStatus.InvalidMessageError,
+          TransactionalEmailMessageStatus.UnsupportedFileTypeError,
+          TransactionalEmailMessageStatus.MaliciousFileError,
+          TransactionalEmailMessageStatus.BlacklistedRecipientError,
+          TransactionalEmailMessageStatus.Accepted,
+          TransactionalEmailMessageStatus.Sent
+        ),
+        allowNull: false,
+      },
+      sent_at: {
+        allowNull: true,
+        type: Sequelize.DataTypes.DATE,
+      },
+      error_code: {
+        allowNull: true,
+        type: Sequelize.DataTypes.STRING(255),
+      },
+      created_at: {
+        allowNull: false,
+        type: Sequelize.DataTypes.DATE,
+      },
+      updated_at: {
+        allowNull: false,
+        type: Sequelize.DataTypes.DATE,
+      },
+    })
+  },
+  down: async (queryInterface, _) => {
+    await Promise.all([
+      queryInterface.dropTable('email_messages_tx'),
+      queryInterface.sequelize.query(
+        'DROP TYPE "enum_email_messages_tx_status";'
```

Comment:
needed for db migrations to be fully reversible (this suggests earlier migrations that create enums without this are not really reversible)

this is because if you drop the table, the enum still persists (weirdly)

Reply:
It's not weird, it's a feature of sql databases - enums, and these are in no way related to the table created. They might be used in table field type declaration, but there's no cascading effect of removing the table on the enum type.

Reply:
On the same note, u should remove the table first before removing the type (rn u're trying to do them at the "same" time which might cause flaky behaviors), because the enum type can't be removed if it's still used in the table declaration.

Reply:
> On the same note, u should remove the table first before removing the type (rn u're trying to do them at the "same" time which might cause flaky behaviors), because the enum type can't be removed if it's still used in the table declaration. 

Ohh you're right

Reply:
> It's not weird, it's a feature of sql databases - enums, and these are in no way related to the table created. They might be used in table field type declaration, but there's no cascading effect of removing the table on the enum type.

Yeah that's fair, I was referring to the fact that our earlier migrations actually don't take this into account (so the down migration doesn't really entirely reverse the up migration)

I was just thinking, is it safe to use the following in the migration (as opposed to enumerating all the enums)?
```
status: {
        type: Sequelize.DataTypes.ENUM(
          ...Object.values(TransactionalEmailMessageStatus)
        ),
        allowNull: false,
      },
```

My thinking is, the shape of `TransactionalEmailMessageStatus` is actually not fixed and if I create more enums in the future, they will actually be created on the first migration right? 🤔 But that makes the migration longer and uglier